File: 122-best-time-to-buy-and-sell-stock-ii/122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
       int sd = 0;
       int bd = 0;
        int n = prices.size();
        int prof = 0;
       for(int i = 1;i<n;i++){
           if(prices[i]>prices[i-1]){
               sd = i;
           }
           else{
               prof += prices[sd]-prices[bd];
               sd = bd = i;
           }
       }
        prof+=prices[sd] - prices[bd];
        return prof;
        
    }
};

File: 122-best-time-to-buy-and-sell-stock-ii/122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int maxProfit(int[] prices) {
         int ossp = 0;
        int obsp = -prices[0];
        for(int i = 1;i<prices.length;i++){
            int nbsp = Math.max(ossp-prices[i], obsp);
            int nssp = Math.max(prices[i]+obsp, ossp);
            obsp = nbsp;
            ossp = nssp;
        }
        return ossp;
    }
}